File: ts/stlc/stlc_test.go
package stlc_test

import (
	"fmt"
	"strings"
	"testing"

	"github.com/jabolopes/bapel/comp"
	"github.com/jabolopes/bapel/ir"
	"github.com/jabolopes/bapel/query"
	"github.com/jabolopes/bapel/tests"
)

func checkUnit(filename string, typecheck bool) (ir.IrUnit, error) {
	querier, err := query.New()
	if err != nil {
		return ir.IrUnit{}, err
	}

	options := comp.TypecheckOptions{
		SkipDefaultContext:      true,
		SkipTermTypechecker:     !typecheck,
		SkipUndefinedTermChecks: true,
	}

	return comp.TypecheckSourceFile(querier, options, filename)
}

func TestInferTerm(t *testing.T) {
	matches, err := tests.Glob("*.in")
	if err != nil {
		t.Fatal(err)
	}

	for _, inFile := range matches {
		wantFile := fmt.Sprintf("%s.out", strings.TrimSuffix(inFile, ".in"))

		sourceFile, err := checkUnit(inFile, false /* typecheck */)
		if err != nil {
			t.Fatalf("in test %s: %v", inFile, err)
		}

		got := fmt.Sprintf("%+s\n", sourceFile)

		diff, err := tests.DiffOutRegen(got, wantFile)
		if err != nil {
			t.Fatalf("in test %s: %v", inFile, err)
		}
		if len(diff) > 0 {
			t.Errorf("in test %s: diff = %s", inFile, diff)
		}
	}
}

func TestTypecheckTerm(t *testing.T) {
	matches, err := tests.Glob("*.in")
	if err != nil {
		t.Fatal(err)
	}

	for _, inFile := range matches {
		if _, err := checkUnit(inFile, true /* typecheck */); err != nil {
			t.Fatalf("in test %s: %v", inFile, err)
		}
	}
}

File: tests/stlc_test.cc
#include "tests/test_util.h"
#include "comp/typecheck_unit.h"
#include "comp/querier.h"
#include "comp/module_finder.h"

#include <string>
#include <vector>

TEST(StlcTest, InferTerm) {
  auto matches = tests::glob("tests/testdata/stlc/*.stlc.in");
  ASSERT_FALSE(matches.empty());

  comp::ModuleFinder finder({}, {{"", "."}});
  comp::Querier querier(finder);

  for (const auto& inFile : matches) {
    ctx.run(inFile, [&](tests::TestContext& sub_ctx) {
      std::string wantFile = tests::replace_extension(inFile, ".out");

      comp::TypecheckOptions options;
      options.skip_default_context = true;
      options.skip_term_typechecker = true;
      options.skip_undefined_term_checks = true;

      ir::IrUnit unit;
      std::string err;
      bool ok = comp::typecheck_source_file(querier, options, inFile, unit, err);

      std::string got;
      if (ok) {
        got = unit.to_bpl_string(true);
      } else {
        got = err + "\n";
      }

      std::string diff_str;
      if (!tests::diff_out_regen(got, wantFile, diff_str)) {
        sub_ctx.add_error("in test " + inFile + ":\n" + diff_str, __FILE__, __LINE__);
      }
    });
  }
}

TEST(StlcTest, TypecheckTerm) {
  auto matches = tests::glob("tests/testdata/stlc/*.stlc.in");
  ASSERT_FALSE(matches.empty());

  comp::ModuleFinder finder({}, {{"", "."}});
  comp::Querier querier(finder);

  for (const auto& inFile : matches) {
    ctx.run(inFile, [&](tests::TestContext& sub_ctx) {
      comp::TypecheckOptions options;
      options.skip_default_context = true;
      options.skip_term_typechecker = false;
      options.skip_undefined_term_checks = true;

      ir::IrUnit unit;
      std::string err;
      bool ok = comp::typecheck_source_file(querier, options, inFile, unit, err);
      if (!ok) {
        sub_ctx.add_error("in test " + inFile + ": " + err, __FILE__, __LINE__);
      }
    });
  }
}
